rust/src/arrow/linalg.rs

```diff
@@ -37,6 +46,113 @@ fn transpose(input: &[f32], dimension: usize) -> Vec<f32> {
     mat
 }
 
+/// A 2-D matrix view on top of Arrow Arrays.
+///
+/// [MatrixView] does not own the data.
+#[derive(Debug)]
+pub struct MatrixView {
+    /// Underneath data array.
+    pub data: Arc<Float32Array>,
+
+    /// The number of
+    pub num_columns: usize,
+
+    /// Is this matrix transposed or not.
+    pub transpose: bool,
+}
+
+impl MatrixView {
+    /// Create a MatrixView from a
+    pub fn new(data: Arc<Float32Array>, num_columns: usize) -> Self {
+        Self {
+            data,
+            num_columns,
+            transpose: false,
+        }
+    }
+
+    /// Number of rows in the matrix
+    pub fn num_rows(&self) -> usize {
+        if self.transpose {
+            self.num_columns
+        } else {
+            self.data.len() / self.num_columns
+        }
+    }
+
+    pub fn num_columns(&self) -> usize {
+        if self.transpose {
+            self.data.len() / self.num_columns
+        } else {
+            self.num_columns
+        }
+    }
+
+    /// (Lazy) transpose of the matrix.
+    ///
+    pub fn transpose(&self) -> Self {
+        Self {
+            data: self.data.clone(),
+            num_columns: self.num_columns,
+            transpose: !self.transpose,
+        }
+    }
+
+    /// Dot multiply
+    pub fn dot(&self, rhs: &Self) -> Result<Self> {
```

Comment:
would it make sense to eventually move all of the distance metrics into this module?

Reply:
That's a good point. We could potentially maintain a `arrow-linalg` someday too :)